common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -6001,8 +5999,6 @@ static int sslSelect(JNIEnv* env, int type, jobject fdObject, AppData* appData,
             timeout_millis = -1;
         }
 
-        CompatibilityCloseMonitor monitor(intFd);
```

Comment:
I did some more static analysis and I think we're mostly OK.  If the socket gets closed by another Conscrypt thread then the emergency pipe takes care of things. If something on the Java side closes the underlying socket, the poll *should* return  (I'll take an AI to test that on Linux and Android, especially for the POLLOUT case!).  

My only slight concern is that the loops in sslSelect() and NativeCrypto_SSL_do_handshake() don't seem to adjust their timeout values after blocking calls, but it's "probably" ok.

Reply:
>  the poll should return

Seems like it does not, see http://r.android.com/879713

For what it's worth, Android once had a similar bug (that one was nondeterministic so it took much longer to find), not sure to what extent the mechanisms of these two bugs overlap. For reference, the fix CL from the previous bug is http://r.android.com/231812

Reply:
Reverting.

Also yes, it looks like this method might be vulnerable to the race described in http://r.android.com/231812 as there is no synchronisation in the code which extracts the file descriptor and no test for it ending up as -1 (which poll will then ignore, potentially waiting forever).  Only affects the case where an SSL socket is closed from Java via a non-BoringSSL method though, otherwise the "emergency" pipe should protect things.